CommonTools/RecoAlgos/plugins/TOFPIDProducer.cc

```diff
@@ -0,0 +1,308 @@
+        
+        //recompute t0 for alternate mass hypotheses
+        double t0_best = t0;
+        
+        float tmtd = tmtdIn[trackref];
+        float pathlength = pathLengthIn[trackref];
+        float magp = pIn[trackref];
+        
+        double gammasq_k = 1. + magp*magp/m_k/m_k;
+        double beta_k = std::sqrt(1.-1./gammasq_k);
+        double t0_k = tmtd - pathlength/beta_k/c;
+        
+        double gammasq_p = 1. + magp*magp/m_p/m_p;
+        double beta_p = std::sqrt(1.-1./gammasq_p);
+        double t0_p = tmtd - pathlength/beta_p/c;
+        
+        double chisqmin = chisqnom;
+        
+        double chisqmin_pi = chisqnom;
+        double chisqmin_k = std::numeric_limits<double>::max();
+        double chisqmin_p = std::numeric_limits<double>::max();
+        //loop through vertices and check for better matches
+        for (const reco::Vertex &vtx : vtxs) {
+          if (!(vtx.tError()>0. && vtx.tError()<vtxMaxSigmaT_)) {
+            continue;
+          }
+          
+          double dz = std::abs(track.dz(vtx.position()));
+          if (dz>=maxDz_) {
+            continue;
+          }
+          
+          double chisqdz = dz*dz*rsigmazsq;
+          
+          double dt_k = std::abs(t0_k - vtx.t());
+          double dtsig_k = dt_k*rsigmat0;
+          double chisq_k = chisqdz + dtsig_k*dtsig_k;
+          
+          if (dtsig_k < maxDtSignificance_ && chisq_k<chisqmin_k) {
+            chisqmin_k = chisq_k;
+          }
+          
+          double dt_p = std::abs(t0_p - vtx.t());
+          double dtsig_p = dt_p*rsigmat0;
+          double chisq_p = chisqdz + dtsig_p*dtsig_p;
+          
+          if (dtsig_p < maxDtSignificance_ && chisq_p<chisqmin_p) {
+            chisqmin_p = chisq_p;
+          }
+          
+          if (dtsig_k < maxDtSignificance_ && chisq_k<chisqmin) {
+            chisqmin = chisq_k;
+            t0_best = t0_k;
+          }
+          if (dtsig_p < maxDtSignificance_ && chisq_p<chisqmin) {
+            chisqmin = chisq_p;
+            t0_best = t0_p;
+          }
+          
+        }
+        
+        //compute PID probabilities
+        //*TODO* deal with heavier nucleons and/or BSM case here?
+        double rawprob_pi = exp(-0.5*chisqmin_pi);
+        double rawprob_k = exp(-0.5*chisqmin_k);
+        double rawprob_p = exp(-0.5*chisqmin_p);
+        
+        double normprob = 1./(rawprob_pi + rawprob_k + rawprob_p);
+        
+        prob_pi = rawprob_pi*normprob;
+        prob_k = rawprob_k*normprob;
+        prob_p = rawprob_p*normprob;
+
+        double prob_heavy = 1.-prob_pi;
+        
+        if (prob_heavy>minProbHeavy_) {
+          t0 = t0_best;
+        }
+
+      }
+      
+    }
+    
+    t0OutRaw.push_back(t0);
+    sigmat0OutRaw.push_back(sigmat0);
+    probPiOutRaw.push_back(prob_pi);
+    probKOutRaw.push_back(prob_k);
+    probPOutRaw.push_back(prob_p);
+  }
+
+  edm::ValueMap<float>::Filler fillert0s(*t0Out);
```

Comment:
code repetition could be reduced by defining a helper function to fill the value maps (see also https://github.com/cms-sw/cmssw/pull/25063/files#r241820150)